hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java

```diff
@@ -1540,8 +1542,14 @@ public List<OzoneFileStatus> listStatus(OmKeyArgs args, boolean recursive,
     countEntries = 0;
     // Convert results in cacheKeyMap to List
     for (OzoneFileStatus fileStatus : cacheKeyMap.values()) {
-      // No need to check if a key is deleted or not here, this is handled
-      // when adding entries to cacheKeyMap from DB.
+      // Here need to check if a key is deleted as cacheKeyMap will contain
+      // deleted entries as well. Adding deleted entries in cacheKeyMap is done
+      // as there is a possible race condition where table cache iterator is
+      // flushed already and isKeyDeleted check may not work as expected
```

Comment:
@sadanand48  Thanks for your review. Pls check the [changes](https://github.com/apache/ozone/pull/5244/files#diff-bde0dade7dd5ddda419499f4f999d25d40fcec1412e0ce809c36ffd1be473f22R1430-R1431) in commit where we do not want to skip deleted keys, `getIteratorForKeyInTableCache` methods gets called by `org.apache.hadoop.ozone.om.KeyManagerImpl#listStatus(org.apache.hadoop.ozone.om.helpers.OmKeyArgs, boolean, java.lang.String, long, java.lang.String, boolean)` where we call cacheIterator on keyTable and then populate cacheKeyMap in `listStatusFindKeyInTableCache` with non-deleted keys and that populated cacheKeyMap is being used in `findKeyInDbWithIterator` method for wrapping the OmKeyInfo object in OzoneFileStatus by having a check if key is not deleted. Now here `isKeyDeleted` was used to check if key is deleted or not. This code can have a race condition where keyTable cacheValue may itself may be null as none of these methods are synchronized as a whole.  Below `omKeyInfoCacheValue` may be null.
```
CacheValue<OmKeyInfo> omKeyInfoCacheValue
        = keyTable.getCacheValue(new CacheKey(key));
```
So in `listStatusFindKeyInTableCache` method itself when iterating cacheIter , we populate cacheKeyMap with deleted keys also.

Reply:
Thanks @devmadhuu for explaining, missed the part where you added the null value. makes sense to avoid the isKeyDeleted  check. We should check other usages of `isKeyDeleted` if it causes similar issues elsewhere.

Maybe the correct way to check isKeyDeleted would be to check the existence of the key in table if no value is present in cache.

Reply:
> Thanks @devmadhuu for explaining, missed the part where you added the null value. makes sense to avoid the isKeyDeleted check. We should check other usages of `isKeyDeleted` if it causes similar issues elsewhere.
> 
> Maybe the correct way to check isKeyDeleted would be to check the existence of the key in table if no value is present in cache.

Thanks @sadanand48 , I have handled the check inside isKeyDeleted method to check the existence of the key in table if no value is present in cache. Few other methods are calling this method.

Reply:
If the problem about cache flush after obtaining the iterator reference of the table is valid (which is the race condition mentioned here in the PR) , I can see this call as problematic.

Let's take example of the caller `createFakeDirIfShould` , it returns a fake dir instance to getFileStatus call if the key exists and if intermediate paths don't. Here it decides whether it should create a fake dir or not based on whether a key beyond the prefix exists in the table. If it exists, we also check if it's deleted based on isKeyDeleted check. Now if it is deleted, we will find the value in cache and fake dir is not created and getfileStatus will return null. But if there is a cache flush, then we won't find any entry in cache and this check will return false and create a fake dir but the key itself doesn't exist.  This can cause getFileStatus to be flaky. If we are not fixing this method, we should probably evaluate its usage and address problems it can cause.
```java
 if (key.startsWith(targetKey)) {
            if (!Objects.equals(key, targetKey)
                && !isKeyDeleted(key, keyTable)) {
              LOG.debug("Fake dir {} required for {}", targetKey, key);
              return createDirectoryKey(keyValue.getValue(), dirKey);
            }
```

Reply:
Yes @sadanand48 , you are right about above method call issue as well, however this has already been identified and being fixed in other PR where getFileStatus call flow is there

Reply:
Cool thanks